### ISSUE TITLE
Improve Issue Experience for .NET VS Code Extensions

### ISSUE THREAD

**Issue by nagilson** [dotnet/vscode-dotnet-runtime#885]
# Context
Currently, we have an `ErrorHandler` that handles all Errors thrown in the VS Code .NET SDK and VS Code .NET Runtime Extension. 

This `ErrorHandler` has an `IssueReporter` that lets customers create a GH issue on our repo by opening the "create an issue url" when they cancel the error window. They often don't understand what will happen when they click to report the error, and they often do not fill in the needed details. So, we get hundreds of non-issues that are unresolvable like this one: https://github.com/dotnet/vscode-dotnet-runtime/issues/871

We would like to not have to close hundreds of blank issues. We cannot remove the privacy statement or auto-fill the issue with the log file seen in the above issue due to privacy concerns. Another problem is that the logs that this message recommends can end up empty, and that should be fixed. I have not investigated why this is the case.

# What to do? 

There are several ways to remedy this.
Here's the first few things:
1) We have no issue template. We should remedy this by adding an issue template to the repo.
2) We should fix the log file issue so they aren't empty. https://github.com/dotnet/vscode-dotnet-runtime/issues/731 Is an example of someone attaching the log file, yet the log file is empty. This is the case for many issues in the repo. We don't know why this happens yet as it hasn't been investigated.
3) We should redirect users to the common issues page and modify that page to link to the create an issue page if their issue isn't found.
https://github.com/dotnet/vscode-dotnet-runtime/blob/main/Documentation/troubleshooting-runtime.md is the page.

I would like to have @baronfel chime in on this before work is started. This is 'low' priority.

cc @marcpopMSFT

I hope this is a good first-issue that is also high-impact. The code for this can be found here: https://github.com/dotnet/vscode-dotnet-runtime/blob/main/vscode-dotnet-runtime-library/src/Utils/IssueReporter.ts

**Comment by nagilson** [comment]
A good proposal should not have A, B, C, D, and instead have the one option. I will arrange something @baronfel 
 @JL03-Yue so we can discuss this and come to a final decision.

**Comment by baronfel** [comment]
Definitely agree on fixing 1. and 2.

For the remainder, We probably have two modes of operation for this extension:
A) we are completely in control of runtime/SDK acquisition, 
B) we are a _client_ of some other extension that's just using us as a means to an end

Our error reporting/user interaction experience should differ based on client type (which could be signalled via the API's of our methods).

All current users are type A - they call us, we throw exceptions and present to users a feedback form. This contract shouldn't change, but we should enhance the granularity of responses the users have options for.  Let's start with type B users and work our way back.

I propose that for Type B clients we _should not throw exceptions_. Instead, we should return a structure like a `Result<'t, 'error>` - a structure that contains a value that is either a success or failure, and the `'error` generic parameter here should be a TypeScript enum/union/whatever that specifies the different kinds of failure we can have.  This would allow clients to detect these different failure types and determine whatever behavior they would like to have.

This would _also_ enable us to have our Type A users to have a richer error experience - we could encode the error data into whatever template we send them to, and perhaps present different options for end users.  That would require us to classify/model the kinds of failures we might have. Stealing from F# a bit here, we might have a structure like the following for errors:

```fsharp
type AcquisitionError = 
| ScriptDownloadFailed // couldn't download the install-scripts
| MissingPrivileges // couldn't elevate/sudo
| NoCompatibleSdkVersions of requestedVersion: string * availableVersions: string list
...
```

So we could model all of these different failure cases and handle them in distinct ways:

ScriptDownloadFailed -> at least tell the user "we couldn't download the scripts from URL, please check connectivity to URL"
MissingPrivileges -> "ensure you have `sudo` permissions, or rerun VSCode as a user that does..."
NoCompatibleSdkVersions -> present a prompt to the user with buttons for the available versions, etc, et

What do you all think?

**Comment by nagilson** [comment]
I align with your idea that exceptions should not be thrown. But I'm not sure about the change with the `Result<>` structure. If we started with it: I love it! But to resolve the above issue, we would have to change the interface for every single API, right? While for functions that don't return something I think it's generally going to be OK, I'm hesitant to change the API still. For those that do, it's pretty big of a change. We could add a separate API layer for it, but I don't really like having a copy API just to throw a new type of error.

There exists the current function `callWithErrorHandling` that checks if `DisableErrorPopups` or `DisplayAllErrorPopups` is set. I wrote a test for it and it didn't work for me... https://github.com/dotnet/vscode-dotnet-runtime/commit/82abdcb7ba1fe11860d56f49c4746eced4a1e6c3

But I think using this existing approach would be a way to not throw messages up when in use with type B clients: https://github.com/dotnet/vscode-dotnet-runtime/blob/82abdcb7ba1fe11860d56f49c4746eced4a1e6c3/vscode-dotnet-runtime-library/src/Utils/ErrorHandler.ts#L15

The error would still be thrown but the popup shouldn't occur. We'd need to make sure that this functionality works though and extend it a bit to be more granular and support more of our errors. We'd create a separate error file or something for all of our exception types and docstring them up all pretty.

Thoughts?

**Comment by nagilson** [comment]
Here's what I think we should do for the third part. It turns out there is already a lot of error events in the eventstreamevents. I think we should redirect users to the common issues page at first and put a note at the beginning of the page telling users to create an issue if they can't find a resolution on the common issues page. This allows us to still get real issues that customers are facing without as many impulsive issues that have absolutely no information attached and make it harder to triage the real issues.

We should expand that page a little bit with the known issues, but let's do that outside of this issue.

**Comment by JL03-Yue** [comment]
Closes as implemented in #1122 and #1123
